rfc/custom-modules-v1.md

```diff
@@ -0,0 +1,706 @@
+
+As powerful as the new module becomes, it is important to move basic and common functionality into the core of the router because building and maintaining custom modules should be a last resort. The router should provide a rich set of features out of the box that cover the most common use cases. Custom modules should be reserved for advanced or highly specific use cases that cannot be achieved with the built-in features of the router. Integration with third-party services, custom authentication, and advanced telemetry are examples of use cases that are well-suited for custom modules.
+
+## Proposal
+
+A developer can implement a custom module by creating a struct that implements one or more of the following interfaces:
+
+- RouterHooks: Provides hooks for the router lifecycle, including request and response handling.
+  - `RouterRequestHook`: Called when a request is made to the router and after all GraphQL information is available.
+  - `RouterResponseHook`: Called before the response is sent to the client.
+  - `RouterErrorHook`: Called when an error occurs during the router lifecycle.
+- SubgraphHooks: Provides hooks for subgraph requests and responses.
+  - `SubgraphRequestHook`: Called when a subgraph request is made.
+  - `SubgraphResponseHook`: Called when a subgraph response is received.
+- ApplicationHooks: Provides hooks for the application lifecycle, including startup, shutdown, and error handling.
+  - `ApplicationStartHook`: Called when the application starts.
+  - `ApplicationStopHook`: Called when the application stops.
+  - `ApplicationErrorHook`: Called when an error occurs during the application lifecycle.
+- AuthenticationHooks: Provides hooks for authentication and authorization logic.
+  - `AuthenticationHook`: Called when a router request is authenticated.
+- AuthorizationHooks: Provides hooks for authorization logic.
+  - `AuthorizationHook`: Called when a router request is authorized.
+- TelemetryHooks: Provides hooks for OpenTelemetry tracing and metrics.
+  - `TelemetrySpanHook`: Called when a span is created.
+  - `TelemetryMetricHook`: Called when a metric is recorded.
+- GraphServerHooks: Provides hooks for the lifecycle of a GraphQL server.
+  - `GraphServerStartHook`: Called when the GraphQL server starts e.g. for the first time or when the schema is updated.
+  - `GraphServerStopHook`: Called when the GraphQL server stops e.g. when the application is shut down or the old schema is replaced.
+- GraphQLOperationHooks: Provides hooks for parsed, normalized, and planned GraphQL operations.
+  - `GraphQLOperationParseHook`: Called when an operation is parsed.
+  - `GraphQLOperationNormalizeHook`: Called when an operation is normalized.
+  - `GraphQLOperationPlanHook`: Called when an operation is planned.
```

Comment:
Will user be able to use GraphQLOperationPlanHook for monitoring the planCache or a separate GraphqlOperationPlanCacheHook is needed? It could be useful for monitoring the cache hit/miss and maybe even modify the caching policy based on their needs.

Reply:
That is a possible way. The current direction is to provide additional information in the context e.g. if a plan was fetched from the cache could be on the context of the `PostGraphQLOperationPlanHook`